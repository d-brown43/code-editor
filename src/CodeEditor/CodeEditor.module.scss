@import '../variables.scss';
@import './code-editor-variables.scss';

.textEditor {
  border: none;
  overflow: auto;
  outline: none;

  -webkit-box-shadow: none;
  -moz-box-shadow: none;
  box-shadow: none;

  resize: none;
  line-height: $line-height;
}

.textEditor,
.container {
  flex: 1;
  background-color: $background-colour;
  color: $foreground-colour;
}

.container {
  display: flex;
  flex-direction: column;
  padding: 40px;
}

.content {
  position: relative;
  flex: 1;
  display: flex;
  flex-direction: column;
}