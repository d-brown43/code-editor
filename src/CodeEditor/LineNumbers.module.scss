@import './code-editor-variables.scss';

.lineNumbers {
  position: absolute;
  margin-left: -16px;
  display: flex;
  flex-direction: column;
  line-height: $line-height;
  letter-spacing: normal;
}
