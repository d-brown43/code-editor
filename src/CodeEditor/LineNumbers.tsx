import * as React from "react";
import styles from './LineNumbers.module.scss';
import {generateRange} from "../helpers";

type LineNumbers = {
    program: string;
};

const LineNumbers = ({program}: LineNumbers) => {
    const lineCount = program.split('\n').length;

    return (
        <div className={styles.lineNumbers}>
            {generateRange(lineCount, 1).map((index) => {
                return (
                    <span key={index}>{index}</span>
                )
            })}
        </div>
    );
};

export default LineNumbers;
