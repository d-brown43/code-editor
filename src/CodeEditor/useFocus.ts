import * as React from "react";
import {RefObject} from "react";

type FocusType = {
    isInitialised: boolean;
    intermediateCopy: string | undefined;
}

export default (textArea: RefObject<HTMLTextAreaElement>) => {
    const reducer = (state: FocusType, action: { type: string, payload?: any }) => {
        switch (action.type) {
            case 'value':
                return {
                    ...state,
                    intermediateCopy: action.payload
                };
            case 'initialised':
                return {
                    ...state,
                    isInitialised: true,
                    intermediateCopy: undefined
                };
            default:
                return state;
        }
    };

    const initialState: FocusType = {
        isInitialised: false,
        intermediateCopy: undefined
    };

    const [{
        isInitialised,
        intermediateCopy
    }, dispatch] = React.useReducer(reducer, initialState);

    React.useEffect(() => {
        if (
            (textArea.current !== null)
            && !intermediateCopy
            && !isInitialised
        ) {
            textArea.current.focus();
            dispatch({type: 'value', payload: textArea.current.value});
            textArea.current.value = '';
        }
    }, [intermediateCopy, isInitialised, textArea]);

    React.useEffect(() => {
        if (
            (textArea.current !== null)
            && intermediateCopy
            && !isInitialised
        ) {
            dispatch({type: 'initialised'});
            textArea.current.value = intermediateCopy;
        }
    }, [isInitialised, intermediateCopy, textArea]);
};
