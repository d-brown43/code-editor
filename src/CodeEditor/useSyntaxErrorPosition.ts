import * as React from "react";
import getCaretCoordinates from "textarea-caret";
import {RefObject} from "react";

export default (
    programError: ProgramError,
    textArea: RefObject<HTMLTextAreaElement>,
    program: string,
    setError: (error: ErrorState) => void
) => {
    React.useEffect(() => {
        if (programError && (textArea.current !== null)) {
            const offset = program.split('\n')
                .filter((ele, index) => index < programError.location.line - 1)
                .reduce((result, line) => (result + 1 + line.length), 0) + programError.location.column;

            const {top, left} = getCaretCoordinates(textArea.current, offset);

            setError({
                x: left,
                y: top,
                error: programError.message
            })
        } else if (!programError) {
            setError({
                x: 0,
                y: 0,
                error: null
            });
        }
    }, [program, programError, setError, textArea]);
};
