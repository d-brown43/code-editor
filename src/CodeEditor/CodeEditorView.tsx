import * as React from "react";
import styles from './CodeEditor.module.scss';
import useTextAreaFocus from './useFocus';
import LineNumbers from "./LineNumbers";
import useSyntaxErrorPosition from "./useSyntaxErrorPosition";

type CodeEditorView = {
    program: string;
    setProgram: (program: string, callback?: (program: string) => void) => void;
    programError: ProgramError;
}

export default React.memo(({program, setProgram, programError}: CodeEditorView) => {
    const textArea = React.useRef<HTMLTextAreaElement>(null);
    const [error, setError] = React.useState<ErrorState>({
        x: 0,
        y: 0,
        error: null
    });

    useTextAreaFocus(textArea);
    useSyntaxErrorPosition(programError, textArea, program, setError);

    return (
        <div className={styles.container}>
            <div className={styles.content}>
                {(error.error === null) ? null : (
                    <span
                        style={{
                            position: 'absolute',
                            top: `calc(${error.y}px + 1em)`,
                            left: error.x,
                            color: 'red'
                        }}
                    >
                        <i className="fas fa-chevron-up" />
                        <span style={{marginLeft: 8}}>
                            {error.error}
                        </span>
                    </span>
                )}
                <LineNumbers program={program} />
                <textarea
                    ref={textArea}
                    className={styles.textEditor}
                    value={program}
                    onChange={(e) => setProgram(e.target.value)}
                />
            </div>
        </div>
    )
});
