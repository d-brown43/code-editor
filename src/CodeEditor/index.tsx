import * as React from "react";
import CodeEditorView from './CodeEditorView';

type CodeEditor = {
    program: string;
    setProgram: (program: string) => void;
    programError: ProgramError;
}

const CodeEditor = ({program, setProgram, programError}: CodeEditor) => (
    <CodeEditorView
        program={program}
        setProgram={setProgram}
        programError={programError}
    />
);

export default CodeEditor;
