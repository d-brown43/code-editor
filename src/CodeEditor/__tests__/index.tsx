import CodeEditor from "../index";
import {shallow} from "enzyme";
import * as React from "react";

it('matches snapshot', () => {
    const codeEditor = shallow(
        <CodeEditor
            program={`console.log('Hello world');`}
            setProgram={jest.fn()}
            programError={false}
        />
    );
    expect(codeEditor).toMatchSnapshot();
});
