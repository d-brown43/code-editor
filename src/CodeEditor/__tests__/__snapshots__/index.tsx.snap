// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Memo()
  program="console.log('Hello world');"
  programError={false}
  setProgram={[MockFunction]}
/>
`;
